lib/protein/processor.rb

```diff
@@ -1,15 +1,27 @@
 module Protein
 class Processor
   class << self
+    DEFAULT_AROUND_PROCESSOR = -> (block) { block.call }
+
     def call(router, request_payload)
+      around_processing = router.config.fetch(:around_processing) do
+        DEFAULT_AROUND_PROCESSOR
+      end
+
+      response = nil
+
       service_name, request_buf = Payload::Request.decode(request_payload)
       service_class = router.resolve_by_name(service_name)
 
-      if service_class.response?
-        process_and_log_call(service_name, service_class, request_buf)
-      else
-        process_and_log_push(service_name, service_class, request_buf)
-      end
+      around_processing.call(-> do
+        response = if service_class.response?
+          process_and_log_call(service_name, service_class, request_buf)
+        else
+          process_and_log_push(service_name, service_class, request_buf)
+        end
+      end, service_name, service_class)
```

Comment:
Why pass the service name and class (which are derivative) and not e.g. `request_payload`?

Reply:
You are right: it's pointless to pass both, and it would make more sense to pass either of those. `request_payload` is not decoded yet (it's done much deeper in the processing, see line#43). I wanted to wrap as much of the processing as possible, but recognizing which call is being processed seems crucial from monitoring perspective.

We could take more event-based (telemetry-like) route and publish event at the end of the processing. We could contain all important data, measure full processing time and take advantage of data that was parsed within current block.